File: ram.jl
filename    = ARGS[1]
mode        = ARGS[2]
codeMode    = true
cur         = Int16(0)
mem         = Dict{Int16, Int16}()
mem_trunc   = Dict{Int16, Int}()
regs        = Dict{Char, Int}()
new_regs    = Array{Char, 1}()
new_mem     = Array{Array{Int16, 1}, 1}()


"Write the value v to an address a in memory."
function writeM(a:: Int16, v:: Int16)
    mem[a] = v
    mode == "1" && push!(new_mem, Int16[a, mem[a]])
end

"Read from address a in memory, if address not present, return 0"
function readM(a:: Int16):: Int16
    get(mem, a, 0)
end

"Write the value v to register a if a is not a digit"
function writeR(a:: Char, v:: Int16)
    !('0' <= a <= '9') && (regs[a] = v)
    mode == "1" && push!(new_regs, a)
end

"Output the value in register a. Registers 0-9 only hold their values."
function readR(a:: Char):: Int16
    '0' <= a <= '9' ?
        readValue(string(a)) :
        get(regs, a, Int16(0))
end

"Return the instruction at address a. Only usable once instructions are loaded into memory."
function readI(a:: Int16):: String
    x = readM(a)
    y = readM(a+Int16(1))
    t = [
        x & 0xff, (x >> 8) & 0xff,
        y & 0xff, (y >> 8) & 0xff
    ]
    String(map(x -> Char(x), t))
end

"Return the value `a` as an Int16. `a` can be hex (i.e., 0x1234) or dec (i.e., 1234)."
function readValue(a:: String):: Int16
    startswith(a, "0x") ?
        parse(Int16, replace(a, "0x"=>""), base=16) :
        parse(Int16, a, base=10)
end
readValue(a:: SubString{String}) = readValue(string(a))


destReg = " "
lastInst = ""
println("Opening $filename, loading into memory.")
open(filename, "r") do file
    global codeMode, cur
    for line in readlines(file)
        if startswith(line, "code: ")
            address = split(replace(line, "code: "=>""))[1]
            cur = readValue(String(address))
            codeMode = true
        elseif startswith(line, "data: ")
            header = split(replace(line, "data: "=>""))
            cur = readValue(String(header[1]))
            trunc = length(header) == 2 ? readValue(header[2]) : 0
            mem_trunc[cur] = Int(trunc)
            codeMode = false
        else
            if codeMode
                code = rpad(strip(split(line, "#")[1]), 4)
                if code != "    "
                    # println(code)
                    writeM(
                        cur,   Int16(code[1]) % Int16(256)
                            + (Int16(code[2]) % Int16(256)) * Int16(256)
                    )
                    writeM(
                        cur+Int16(1), Int16(code[3]) % Int16(256)
                            + (Int16(code[4]) % Int16(256)) * Int16(256)
                    )
                    if mode == "4"
                        errString = "Hazard: $lastInst, $code"
                        if code[1] in "BbEe<l>g"  # branches
                            destReg in "$(code[2])" && println(errString)
                            destReg = " "
                            lastInst = code
                        elseif code[1] in "+-*/%"  # maths
                            destReg in "$(code[2])$(code[3])" && println(errString)
                            destReg = string(code[4])
                            lastInst = code
                        elseif code[1] in "JI"  # jumps/immediates
                            destReg = string(code[2])
                            lastInst = code
                        elseif '!' == code[1]  # syscalls
                            destReg in "$(code[2])" && println(errString)
                            destReg = string(code[2])
                            lastInst = code
                        elseif 'L' == code[1]  # loads
                            destReg in "$(code[2])" && println(errString)
                            destReg = string(code[3])
                            lastInst = code
                        elseif 'R' == code[1]  # returns
                            destReg in "$(code[2])" && println(errString)
                            destReg = " "
                            lastInst = code
                        end
                    end
                    cur += Int16(2)
                end
            else
                data = map(
                    x -> readValue(String(x)),
                    split(strip(split(line, "#")[1]))
                )
                for d in data
                    writeM(cur, d)
                    cur += Int16(1)
                end
            end
        end
    end
end


# writeR('P', 0)
halt = false
println("Executing memory.")
while !halt
    global new_regs, new_mem, halt
    inst = readI(readR('P'))
    # println(inst)
    c = inst[1]
    inst = lstrip(inst, c)

    if c == 'L'
        r = inst[1]
        d = inst[2]
        a = readR(r)
        v = readM(a)
        writeR(d, v)
        writeR('P', readR('P') + Int16(2))

    elseif c == 'S'
        s = inst[1]
        w = inst[2]
        a = readR(w)
        v = readR(s)
        writeM(a, v)
        writeR('P', readR('P') + Int16(2))

    # Mathematical operands
    elseif c == '+'
        s1 = inst[1]
        s2 = inst[2]
        d  = inst[3]
        v1 = readR(s1)
        v2 = readR(s2)
        v  = v1 + v2
        writeR(d, v)
        writeR('P', readR('P') + Int16(2))

    elseif c == '-'
        s1 = inst[1]
        s2 = inst[2]
        d = inst[3]
        v1 = readR(s1)
        v2 = readR(s2)
        v = v1 - v2
        writeR(d, v)
        writeR('P', readR('P') + Int16(2))

    elseif c == '*'
        s1 = inst[1]
        s2 = inst[2]
        d = inst[3]
        v1 = readR(s1)
        v2 = readR(s2)
        v = v1 * v2
        writeR(d, v)
        writeR('P', readR('P') + Int16(2))

    elseif c == '/'
        s1 = inst[1]
        s2 = inst[2]
        d = inst[3]
        v1 = readR(s1)
        v2 = readR(s2)
        v = v1 / v2
        writeR(d, v)
        writeR('P', readR('P') + Int16(2))

    elseif c == '%'
        s1 = inst[1]
        s2 = inst[2]
        d = inst[3]
        v1 = readR(s1)
        v2 = readR(s2)
        v = v1 % v2
        writeR(d, v)
        writeR('P', readR('P') + Int16(2))

    # Branch operands
    elseif c == 'B'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) != 0 ?
            writeR('P', readR('P') + readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == 'b'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) != 0 ?
            writeR('P', readR('P') - readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == 'E'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) == 0 ?
            writeR('P', readR('P') + readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == 'e'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) == 0 ?
            writeR('P', readR('P') - readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == '<'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) < 0 ?
            writeR('P', readR('P') + readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == 'l'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) < 0 ?
            writeR('P', readR('P') - readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == '>'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) > 0 ?
            writeR('P', readR('P') + readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    elseif c == 'g'
        s = inst[1]
        hh_hl = inst[2] * inst[3]
        readR(s) > 0 ?
            writeR('P', readR('P') - readValue("0x" * hh_hl) * Int16(2)) :
            writeR('P', readR('P') + Int16(2))

    # Return and halt
    elseif c == 'R'
        writeR('P', readR(inst[1]))

    elseif c == 'H'
        halt = true

    # Jump and Imm
    elseif c == 'J'
        imm = readValue("0x" * readI(readR('P') + Int16(2)))
        writeR(inst[1], readR('P') + Int16(4))
        writeR('P', imm)

    elseif c == 'I'
        imm = readValue("0x" * readI(readR('P') + Int16(2)))
        writeR(inst[1], imm)
        writeR('P', readR('P') + Int16(4))

    # Syscall
    elseif c == '!'
        m = inst[1]
        a = readR(m)
        hh_hl = inst[2] * inst[3]
        if hh_hl == "01"
            while readM(a) != 0
                print(Char(readM(a)))
                a += Int16(1)
            end
        elseif hh_hl == "02"
            writeR(m, readValue(readline(stdin)))
        end
        writeR('P', readR('P') + Int16(2))
    else
        @error("Illegal instruction!")
        break
    end

    ms = !isempty(new_mem) ? ", M: " * string(new_mem) : ""
    rs = !isempty(new_regs) ? ", R: " * string(new_regs) : ""
    if mode == "1" println("$(c * inst)" * rs * ms)
    elseif !(mode in ["2", "3", "4"]) error("Not a valid mode!")
    end

    new_regs = Array{Char, 1}()
    new_mem  = Array{Array{Int16, 1}, 1}()
end  # execution block

# println(regs)
# println(mem_trunc)
if mode == "2"
    # output the values in each data section
    for d in sort(collect(mem_trunc))
        a = d[1]
        print("Address: $(d[1])\nData:    ")
        if d[2] != 0
            for i = 1:d[2]
                print(readM(a))
                a += Int16(1)
            end
            print('\n')
        end
    end
end

File: ram.scala
import scala.io.Source
//TODO or not TODO that is the question
val filename   = args(0)
val mode       = args(1)
var codeMode   = true
var cur        = 0
val mask       = 65535
var mem        = collection.mutable.Map[Int, Int]()
var mem_trunc  = collection.mutable.Map[Int, Int]()
var regs       = collection.mutable.Map[Char, Int]()
var new_regs   = List[Char]()
var new_mem    = List[(Int,Int)]()
var hazard     = List('0', '0', 0) // (reg1, reg2, mem_loc)


def writeM(a: Int, v: Int) = {
    // write the value v to address a in memory, mod mask
    mem(a) = v & mask
    if (mode == "1") new_mem = (a, mem(a))::new_mem
}

def readM(a: Int) = {
    if (!mem.contains(a)) 0
    else mem(a)
}

def writeR(a: Char, v: Int) = {
    // only write to memory if the address is not in [0, 9]
    if (!('0' <= a && a <= '9')) regs(a) = v
    if (mode == "1") new_regs = a::new_regs
}

def readR(a: Char):Int = {
    if('0' <= a && a <= '9'){
        readValue(a.toString)
    } else if (regs.contains(a)) {
        regs(a)
    } else 0
}

def readI(a: Int) = {
    val x = readM(a)
    val y = readM(a+1)
    Array(
        x & 0xff, (x >> 8) & 0xff,
        y & 0xff, (y >> 8) & 0xff
    ).map(_.toChar).mkString("")
}

def readValue(a: String) = {
    if (a.startsWith("0x"))
        Integer.parseInt(a.stripPrefix("0x"), 16)
    else
        Integer.parseInt(a, 10)
}


var stop = false
// for (line <- Source.fromFile(filename).getLines) {
val lines = Source.fromFile(filename).getLines
var writeSeen = false
while (lines.hasNext && !stop) {
    val line = lines.next()
    if (line.startsWith("code: ")) {
        val address = line.stripPrefix("code: ").split(" ")(0)
        cur = readValue(address)
        codeMode = true
    } else if (line.startsWith("data: ")) {
        val n = line.stripPrefix("data: ").split(" ")
        cur = readValue(n(0))  // memory address
        val t = if (n.length == 2) readValue(n(1)) else 0  // store the truncate value in a mapping, if it exists
        mem_trunc(cur) = t
        codeMode = false
    } else {
        if (codeMode) {
            val code = line.split("#")(0).trim().padTo(4,' ')
            if (code != "    ") {
                writeM(cur,   code(0).toInt % 256
                            +(code(1).toInt % 256)*256)
                writeM(cur+1, code(2).toInt % 256
                            +(code(3).toInt % 256)*256)
                if (mode == "4") {
                    if (!writeSeen) code(0) match {
                        case 'B' if (code(1) != 0) => { hazard = List('P', '0', cur); writeSeen = true }
                        case 'b' if (code(1) != 0) => { hazard = List('P', '0', cur); writeSeen = true }
                        case 'E' if (code(1) == 0) => { hazard = List('P', '0', cur); writeSeen = true }
                        case 'e' if (code(1) == 0) => { hazard = List('P', '0', cur); writeSeen = true }
                        case '<' if (code(1) < 0)  => { hazard = List('P', '0', cur); writeSeen = true }
                        case 'l' if (code(1) < 0)  => { hazard = List('P', '0', cur); writeSeen = true }
                        case '>' if (code(1) > 0)  => { hazard = List('P', '0', cur); writeSeen = true }
                        case 'g' if (code(1) > 0)  => { hazard = List('P', '0', cur); writeSeen = true }
                        case 'L' => { hazard = List(code(2), '0', cur); writeSeen = true }
                        case '+' => { hazard = List(code(3), '0', cur); writeSeen = true }
                        case '-' => { hazard = List(code(3), '0', cur); writeSeen = true }
                        case '*' => { hazard = List(code(3), '0', cur); writeSeen = true }
                        case '/' => { hazard = List(code(3), '0', cur); writeSeen = true }
                        case '%' => { hazard = List(code(3), '0', cur); writeSeen = true }
                        case 'J' => { hazard = List(code(1), 'P', cur); writeSeen = true }
                        case 'I' => { hazard = List(code(1), 'P', cur); writeSeen = true }
                        case '!' => { hazard = List(code(1), '0', cur); writeSeen = true }
                        case 'R' => { hazard = List('P', '0', cur); writeSeen = true }
                        case  _  => // do nothing
                    } else {
                        val s = f"Read-write hazard between addresses ${hazard(2)} and ${cur}, instructions ${readI(hazard(2))} and ${code}."
                        code(0) match {
                            case '+' => if (hazard.contains(code(1)) || hazard.contains(code(2)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case '-' => if (hazard.contains(code(1)) || hazard.contains(code(2)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case '*' => if (hazard.contains(code(1)) || hazard.contains(code(2)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case '/' => if (hazard.contains(code(1)) || hazard.contains(code(2)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case '%' => if (hazard.contains(code(1)) || hazard.contains(code(2)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case 'L' => if (hazard.contains(code(1)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case 'S' => if (hazard.contains(code(2)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case '!' => if (hazard.contains(code(1)) || hazard.contains('P')) { println(s); writeSeen = false }
                            case 'R' => if (hazard.contains(code(1))) { println(s); writeSeen = false }
                            case 'B' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'b' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'E' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'e' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case '<' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'l' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case '>' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'g' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'J' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case 'I' => if (hazard.contains('P')) { println(s); writeSeen = false }
                            case _   => writeSeen = false
                        }
                    }
                }
                cur += 2
            }
        } else {
            val data = line.split("#")(0).trim()
                        .split(" ").filter(_ != "")
                        .map(readValue)
            for (d <- data) {
                writeM(cur, d)
                cur += 1
            }
        }
    }
}


writeR('P', 0)
var inst = ""
var halt = false
do {
    inst = readI(readR('P'))
    inst(0) match{
        // Load and Store
        case 'L' => val r = inst(1)
                    val d = inst(2)
                    val a = readR(r)
                    val v = readM(a)
                    writeR(d, v)
                    writeR('P', readR('P') + 2)
        case 'S' => val s = inst(1)
                    val w = inst(2)
                    val a = readR(w)
                    val v = readR(s)
                    writeM(a, v)
                    writeR('P', readR('P') + 2)
        // Mathematical Operands
        case '+' => val s1 = inst(1)
                    val s2 = inst(2)
                    val d = inst(3)
                    val v1 = readR(s1)
                    val v2 = readR(s2)
                    val v = v1 + v2
                    writeR(d, v)
                    writeR('P', readR('P') + 2)
        case '-' => val s1 = inst(1)
                    val s2 = inst(2)
                    val d = inst(3)
                    val v1 = readR(s1)
                    val v2 = readR(s2)
                    val v = v1 - v2
                    writeR(d, v)
                    writeR('P', readR('P') + 2)
        case '*' => val s1 = inst(1)
                    val s2 = inst(2)
                    val d = inst(3)
                    val v1 = readR(s1)
                    val v2 = readR(s2)
                    val v = v1 * v2
                    writeR(d, v)
                    writeR('P', readR('P') + 2)
        case '/' => val s1 = inst(1)
                    val s2 = inst(2)
                    val d = inst(3)
                    val v1 = readR(s1)
                    val v2 = readR(s2)
                    val v = v1 / v2
                    writeR(d, v)
                    writeR('P', readR('P') + 2)
        case '%' => val s1 = inst(1)
                    val s2 = inst(2)
                    val d = inst(3)
                    val v1 = readR(s1)
                    val v2 = readR(s2)
                    val v = v1 % v2
                    writeR(d, v)
                    writeR('P', readR('P') + 2)
        // Branch Operands
        case 'B' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if (readR(s) != 0) {
                        writeR('P', readR('P') + readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case 'b' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) != 0) {
                        writeR('P', readR('P') - readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case 'E' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) == 0) {
                        writeR('P', readR('P') + readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case 'e' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) == 0) {
                        writeR('P', readR('P') - readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case '<' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) < 0) {
                        writeR('P', readR('P') + readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case 'l' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) < 0) {
                        writeR('P', readR('P') - readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case '>' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) > 0) {
                        writeR('P', readR('P') + readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        case 'g' => val s = inst(1)
                    val hh_hl = inst(2).toString + inst(3).toString
                    if(readR(s) > 0) {
                        writeR('P', readR('P') - readValue("0x" + hh_hl) * 2)
                    } else {
                        writeR('P', readR('P') + 2)
                    }
        // Return
        case 'R' => val s = inst(1)
                    writeR('P', readR(s))
        case 'H' => halt = true
        // Jump and Imm
        case 'J' => val imm = readValue("0x" + readI(readR('P') + 2))
                    writeR(inst(1), readR('P') + 4)
                    writeR('P', imm)
        case 'I' => val imm = readValue("0x" + readI(readR('P') + 2))
                    writeR(inst(1), imm)
                    writeR('P', readR('P') + 4)
        // Syscall
        case '!' => val m = inst(1)
                    var a = readR(m)
                    val hh_hl = inst(2).toString + inst(3).toString
                    hh_hl match {
                        case "01" => {
                            while (readM(a) != 0) {
                                print(readM(a).toChar)
                                a += 1
                            }
                        }
                        case "02" => {
                            val scanner = new java.util.Scanner(System.in)
                            writeR(m, readValue(scanner.nextLine()))
                        }
                    }
                    writeR('P', readR('P') + 2)
        case _ => throw new IllegalArgumentException("Instruction not found.")
    }

    mode match {
        case "1" => println(f"I: $inst, R: ${new_regs.mkString("[", ", ", "]")}, M: ${new_mem.mkString("[", ", ", "]")}")
                /*
                Output a trace of execution that includes on each line the instruction executed,
                the new values of any registers changed by that instruction,
                and the address and value of any changed memory locations.
                */
        case "2" =>// is below
                /*
                Execute without tracing (faster) and output the values in each data section at the end of execution.
                Use the second value in each the data section header as the number of memory locations to display at the end.
                */
        case "3" => // perform syscall operands - if not mode 3, then don't do any syscalls

        case "4" =>
                /*
                Analyse the program by outputting all the read-after-write hazards in the program.
                For each one, output the addresses and pair of instructions.
                */
        case _ => throw new IllegalArgumentException("Not a valid mode.")
    }

    new_regs = List[Char]()
    new_mem  = List[(Int,Int)]()

} while (!halt)


if (mode == "2") {
    // output the values in each data section
    // truncated by the second number in the data header, if there is one
    for (d <- mem_trunc.toList.filter(_._2 != 0)) {
        var a = d._1
        print(f"Address: ${d._1}\nData:    ")
        if (d._2 != 0) {
            // print the data starting at a until d._2 is reached
            for (i <- 1 to d._2) {
                print(readM(a))
                a += 1
            }
            print('\n')
        }
    }
}